### DISCUSSION TITLE
NRF low power mode (based on RTC)

### DISCUSSION THREAD

**Discussion by marcel-cd** [contiki-ng/contiki-ng#2860] [Q&A] [answered]
Dear Contiki-NG Team,

The Nordic (nrf) implementation in Contiki-NG uses the Timer (HFXO) for the rtimer. This results in 0.5mA current draw, even during sleep. I made in my contiki-ng fork a migration to the RTC, which is based on the LFXO, and is much better suited for low power devices. The switch between TIMER or RTC can be done with a config-flag in project-conf.h, but it would be cleaner, if I remove this choice, but then a LF XO (32.768kHz) would be mandatory.

I can clean up my code and do a Pull Request, but it's a lot of code change....
What do you think? Do you have time to review bigger PRs? (I cant split this)

Marcel

**Comment by Yagoor** [comment, marked as answer]
Hi Marcel,

It has been a while since I did the nrf port but I think the HFXO had to be used in the rtimer because the radio driver uses the TIMER0 for time stamping. 
Since you have a working solution, I suppose you faced this issue. If you found a good solution for it, I think it would be a great contribution to use the LFXO for the lower power consumption. 

I agree, it does not have any value to have a flag to enable the use of TIMER or RTC. The RTC should be the default. 
I don’t have time to test it, but I’m available to review the code. 

Regards,

Yago

**Reply by marcel-cd** [reply]
Hi Yago,

Thank you for taking your time reviewing the code, I will prepare a PR the next days. 
, Marcel

**Reply by pjonsson** [reply]
As a general comment to make life easier for Yagoor: try to resist changing broken whitespace and similar things that are not related to the low-power mode in the PR. (I'm not arguing against things being wrong somewhere in the Contiki-NG code base, just stating that it's much harder to review a PR consisting of many things.)

I assume the PR will only touch nrf-platform specific files, so you can use C11 freely, including declaring variables when they are used.

**Reply by marcel-cd** [reply]
I've tried to reduce the PR to the very minimum, hope this helps to review the code. The PR touches only the nrf-platform. The result is impressive, When I use a 4 second timeslot, the average current is not more than 25uA.
(I've deactivated in my testproject UARTE and GPIOTE, which would take another 25 uA in sleep)
